openfpga_flow/scripts/run_fpga_task.py

```diff
@@ -263,6 +265,10 @@ def generate_each_task_actions(taskname):
                                                        fallback="top")
         CurrBenchPara["ys_script"] = SynthSection.get(bech_name+"_yosys",
                                                       fallback=ys_for_task_common)
+        CurrBenchPara["yosys_mode"] = SynthSection.get(bech_name+"_yosys_mode",
```

Comment:
This is too restricted and only works for QuickLogic device. OpenFPGA's task run script should not be biased to any devices. We should think about something such as enabling parameter definition in Yosys scripts, just like how we support openfpga shell script.

Reply:
I need to dynamically choose the value in yosys script for yosys_family and yosys_mode. Please elaborate how can it be done by "enabling parameter definition in yosys script". 
Other option I can think of getting these values from environment variables but then what is the right to set these environment variables in OpenFPGA flow.

Reply:
Attn @tangxifan : Any comment on my above reply.

Reply:
Sorry. I miss that one. The idea is like this: 

https://github.com/lnis-uofu/OpenFPGA/blob/e6d1ac4a588363c24f051bf417bcd932a9c66f1f/openfpga_flow/openfpga_shell_scripts/fix_device_route_chan_width_example_script.openfpga#L3

Users can define variables in the script, such as ${OPENFPGA_VPR_DEVICE_LAYOUT}, and the variables can be defined in the task configuration file.

https://github.com/lnis-uofu/OpenFPGA/blob/e6d1ac4a588363c24f051bf417bcd932a9c66f1f/openfpga_flow/tasks/fpga_bitstream/generate_bitstream/device_96x96/config/task.conf#L24

Reply:
Looks like this requires updates to some scripts as I am thinking if it can be supported in this manner:

task.conf:
[SYNTHESIS_PARAM]
yosys_mode = no_adder

qlf_yosys.ys:
synth_quicklogic -blif ${OUTPUT_BLIF} -family ${YOSYS_FAMILY} -top ${TOP_MODULE} ${YOSYS_MODE}

Reply:
@ganeshgore Can you provide some advice?

Reply:
@lnsharma Why cant we just create a separate yosys script with `YOSYS_FAMILY` and `YOSYS_MODE` hard coded

Reply:
This is what I did initially but in general the feedback is to take their values dynamically. Even to me this sounds a better option to take values dynamically.

Reply:
So when you say dynamically, Do you mean in the same task you want to run some benchmarks with Adder mode and other with others without it?

Reply:
I mean to use same ys script to run for adder/no_adder mode, for different families. The choice of family or mode should be specified in task conf file.

Reply:
Can you reset `run_fpga_flow.py`  file to master and add following lines of code just before line 491

This will allow you to define variable in `SCRIPT_PARAM` like you did above 
```
[SYNTHESIS_PARAM]
YOSYS_FAMILY=ql_k4n8
```
and refer in the `*.ys` script as ${YOSYS_FAMILY}

```
for indx in range(0, len(OpenFPGAArgs), 2):
    tmpVar = OpenFPGAArgs[indx][2:].upper()
    ys_params[tmpVar] = OpenFPGAArgs[indx+1]
```

https://github.com/lnis-uofu/OpenFPGA/blob/9f7d03cd059d71a7c3362818bc6020dca388ca7c/openfpga_flow/scripts/run_fpga_flow.py#L491

Reply:
@ganeshgore : please review my changes and let me know if these are fine or you have any comments for improvement.

Reply:
Hi @ganeshgore 

I've run the **run_fpga_flow.py** script with arguments generated by run_fpga_task and got **_IndexError: list index out of range_** error for the following line
https://github.com/lnis-uofu/OpenFPGA/blob/master/openfpga_flow/scripts/run_fpga_flow.py#L591

```
for indx in range(0, len(OpenFPGAArgs), 2):
    tmpVar = OpenFPGAArgs[indx][2:].upper()
    ys_params[tmpVar] = OpenFPGAArgs[indx+1]
```

Is **OpenFPGAArgs** supposed to be populated with an even number of arguments? 
Because the odd length of **OpenFPGAArgs** will give IndexError. 
I'm new to OpenFPGA flow. Just making sure I've configured it properly.

Thanks

Reply:
@asrafrobin 
Yes, that is correct. Because we enforced Keyword based argument and don't support positional argument to this script.
In which case there should be even number of arguments.

Reply:
@ganeshgore 

The odd number (_19 in my case_) is coming because **--yosys_param** argument followed by two input values as follows
`--yosys_param -family ql_k4n8`

So the length of OpenFPGAArgs increase by 3, which ultimately leads to an odd length of OpenFPGAArgs.